lib/rubocop/rspec/top_level_group.rb

```diff
@@ -10,28 +10,31 @@ module TopLevelGroup
       def_node_matcher :example_or_shared_group?,
                        (ExampleGroups::ALL + SharedGroups::ALL).block_pattern
 
-      def on_block(node)
-        return unless respond_to?(:on_top_level_group)
-        return unless top_level_group?(node)
+      def on_new_investigation
+        return unless root_node
 
-        on_top_level_group(node)
+        top_level_groups.each do |node|
+          example_group?(node, &method(:on_top_level_example_group))
+          on_top_level_group(node)
+        end
       end
 
-      def single_top_level_group?
-        top_level_groups.one?
+      def top_level_groups
+        @top_level_groups ||=
```

Comment:
I was following [this note](https://github.com/rubocop-hq/rubocop-rspec/pull/932#discussion_r439636213):

> Right now a new cop instance is created for every file

There might have been some changes after the doc has been introduced.

As a quick experiment, I've set a breakpoint in random cop's `initialize` and ran `rubocop`:
```
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
```

@marcandre Can you please shed light on this? Isn't this doc misleading?

I used to think it would be quite hard to reuse cop instances since depending on the directory they may receive different config (due to directory-local `.rubocop.yml`).

Reply:
The [cop mentioned in the referenced discussion, `SurroundingSpace`](https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/mixin/surrounding_space.rb#L59) does reset its state:
```ruby
      def on_new_investigation
        @token_table = nil
```

Reply:
What's even more strange:
```ruby
        def on_new_investigation
          print "N"
          super
        end

        def initialize(*args)
          print "#"
          super
        end
```
outputs:
```
##.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#C#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#.
```

🤔

Reply:
@pirj Currently, cops being persisted [is opt-in](https://docs.rubocop.org/rubocop/0.88/v1_upgrade_notes.html#cop-persistence). That being said, I'm wondering if a cop that implements `on_new_investigation` shouldn't be persisted by default.
BTW, different config => new cop, but that's a very rare setup. 
It looks like the cop is initialized a bunch of times before doing the actual run, that seems like a bug somewhere, I'll check it out.

Reply:
I'm looking at default cop's initializer and struggling to understand if there are any performance/memory usage improvements of reusing cop instances:
```ruby
      def initialize(config = nil, options = nil)
        @config = config || Config.new
        @options = options || { debug: false }
        reset_investigation
      end

      def reset_investigation
        @currently_disabled_lines = @current_offenses = @processed_source = @current_corrector = nil
      end
```

It would make sense if the expensive `Config.new` would be called on each extra cop initialization, but it isn't:
```ruby
      # Called between investigations
      def ready
        return self if self.class.support_multiple_source?

        self.class.new(@config, @options) # <== `@config` is reused
      end
```

Reply:
Please don't get me wrong, I don't really mind (as a performance-last person) to remove memoization here. We're only calling this `top_level_groups` method once per top-level example group. And those who have more than one, probably deserve to slightly suffer anyway.

I'm more concerned about the confusion that this reuse might deal if once this option will become opt-out rather than opt-in. Cop specs won't be bringing memoization to the light, will they? With all numerous well-unknown RuboCop extensions this can have quite an impact.

Reply:
Nothing major indeed. I wanted to speedup the callback cache building in `Commissionner`, but it could be done another way (per class instead of per instance). Some cops *want to* persist, though (there was an issue on that), which is why it's possible to do so.

Reply:
Also, speaking of per-directory configs. In theory, it's allowed, and I don't see a reason to disrespect this option.

In the practice of working with majestic monoliths composed of Rails engines, in-repo gems and libraries in git submodules it's totally not uncommon to have different styles and different configuration, at least in some directories.

Reply:
> Some cops want to persist

Got it 👍 That's reasonable, I can imagine some use cases for that.

However, does it make sense to keep this notice for the rest of the cops that don't want and won't persist by default? https://github.com/rubocop-hq/rubocop/blob/6b36c4490c2d7c41349e4f71721bf0cf91920578/docs/modules/ROOT/pages/v1_upgrade_notes.adoc#if-your-class-uses-instance-variables

Reply:
> Cop specs won't be bringing memoization to the light, will they? With all numerous well-unknown RuboCop extensions this can have quite an impact.

Yes, that was my thinking too.

> I don't see a reason to disrespect this option.

Absolutely :-) 

> However, does it make sense to keep this notice for the rest of the cops

You mean the recommendation to clear the data with `on_new_investigation`? I guess it's not necessary. I should also be clearer reword the "will be called on different if" to "will not be called unless"...

Reply:
I would remove the whole "If your class uses instance variables" section, as it's more confusing than helpful considering the current "one source" - "one cop instance" behaviour.

Even for those cops that need to be persisted, I can think of ~global variables~ class variables to persist the data between the runs.

By the way, there's one more thing that can impact those persisted cops - `--parallel`. Do we parallelize by cops or by source?

Reply:
=> https://github.com/rubocop-hq/rubocop/pull/8440

Reply:
> Even for those cops that need to be persisted, I can think of class variables to persist the data between the runs.

Not quite... See https://github.com/rubocop-hq/rubocop/issues/7968 if interested

> By the way, there's one more thing that can impact those persisted cops - --parallel. Do we parallelize by cops or by source?

I imagine it's by source, but we could segregate those cops that need it; I suspect they simply run them without `--parallel` though.

Reply:
@bquorning As a bottom-line: I wouldn't worry about `@top_level_groups ||=` used here unless any of the cops would decide to opt-in for reusage *and* use `TopLevelGroup` which I highly doubt.